contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,94 @@
+# Apply v2
+
+## Background
+
+`kubectl apply` reads a file or set of files, and updates the cluster state based off the file contents.
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+Reducing the code complexity and improving the correctness will also make it simpler to move this logic
+to the server as part of an API in the future, while also addressing the immediate needs of the client.
+
+## Changes
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
```

Comment:
We would version it by adding new content types, which I believe we also do with proto encodings.

But discoverability of supported versions is a good point.

Reply:
Re. discoverability: It's possible to specify supported content types for each verb X resource via OpenAPI

Reply:
Please add something about versioning the patch API in the future.